janus_server/src/metrics.rs

```diff
@@ -0,0 +1,178 @@
+//! Collection and exporting of application-level metrics for Janus.
+
+use opentelemetry::{
+    metrics::Descriptor,
+    sdk::export::metrics::{Aggregator, AggregatorSelector},
+};
+use serde::{Deserialize, Serialize};
+use std::{collections::HashMap, net::AddrParseError, sync::Arc};
+
+/// Errors from initializing metrics provider, registry, and exporter.
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("bad IP address: {0}")]
+    IpAddress(#[from] AddrParseError),
+    #[error(transparent)]
+    OpenTelemetry(#[from] opentelemetry::metrics::MetricsError),
+    #[cfg(feature = "otlp")]
+    #[error(transparent)]
+    TonicMetadataKey(#[from] tonic::metadata::errors::InvalidMetadataKey),
+    #[cfg(feature = "otlp")]
+    #[error(transparent)]
+    TonicMetadataValue(#[from] tonic::metadata::errors::InvalidMetadataValue),
+    #[error("{0}")]
+    Other(&'static str),
+}
+
+/// Configuration for collection/exporting of application-level metrics.
+#[derive(Clone, Debug, PartialEq, Eq, Default, Serialize, Deserialize)]
+pub struct MetricsConfiguration {
+    /// Configuration for OpenTelemetry metrics, with a choice of exporters.
+    #[serde(default)]
+    pub exporter: Option<MetricsExporterConfiguration>,
+}
+
+/// Selection of an exporter for OpenTelemetry metrics.
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub enum MetricsExporterConfiguration {
+    #[serde(rename = "prometheus")]
+    Prometheus,
+    #[serde(rename = "otlp")]
+    Otlp(OtlpExporterConfiguration),
+}
+
+/// Configuration options specific to the OpenTelemetry OTLP metrics exporter.
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct OtlpExporterConfiguration {
+    /// gRPC endpoint for OTLP exporter.
+    pub endpoint: String,
+    /// Additional metadata/HTTP headers to be sent with OTLP requests.
+    #[serde(default)]
+    pub metadata: HashMap<String, String>,
+}
+
+/// Choice of OpenTelemetry metrics exporter implementation.
+pub enum MetricsExporterHandle {
+    #[cfg(feature = "prometheus")]
+    Prometheus(
+        opentelemetry_prometheus::PrometheusExporter,
+        tokio::task::JoinHandle<()>,
+    ),
+    #[cfg(feature = "otlp")]
+    Otlp(opentelemetry::sdk::metrics::controllers::PushController),
+    Noop,
+}
+
+#[derive(Debug)]
+struct CustomAggregatorSelector;
+
+impl AggregatorSelector for CustomAggregatorSelector {
+    fn aggregator_for(&self, descriptor: &Descriptor) -> Option<Arc<dyn Aggregator + Send + Sync>> {
+        match descriptor.instrument_kind() {
+            opentelemetry::metrics::InstrumentKind::ValueRecorder => {
+                // The following boundaries are copied from the Ruby and Go Prometheus clients.
+                // Buckets could be specialized per-metric by matching on `descriptor.name()`.
+                let boundaries = &[
```

Comment:
If we create a metric with these boundaries, will we be able to switch later or are we "locked in" for that metric? I'm sure these are reasonable default buckets but the right buckets for a given metric are going to be heavily metric-dependent.